### DISCUSSION TITLE
Faster pane navigation in tmux

### DISCUSSION THREAD

**Discussion by b0o** [mrjones2014/smart-splits.nvim#154] [Show and tell]
The recommended tmux configuration for navigating panes is slower than it needs to be, because it's checking if a pane contains vim at the moment you press the keybinding to switch panes. This leads to a noticeable delay.

I wrote a script for tmux that runs in the background and periodically checks which panes contain vim and sets a pane-local option. Then, when you hit the keybinding to switch panes, tmux only needs to check for this option to determine what to do, rather than shelling out to `ps` and `grep`. With this approach, there is virtually no lag when switching panes.

You don't need to change your vim-tmux-navigator configuration at all to use it.

The script: https://github.com/b0o/tmux-conf/blob/main/navi.tmux

To install:

0. Save the script somewhere
1. Add this to the end of your `tmux.conf`: `run-shell "$XDG_CONFIG_HOME/tmux/navi.tmux"` (change the path to wherever you put the script)
2. Change your pane navigation bindings:

```
bind-key -n C-h if -F "#{@navi-state}" "send-keys C-h" "select-pane -LZ"
bind-key -n C-j if -F "#{@navi-state}" "send-keys C-j" "select-pane -DZ"
bind-key -n C-k if -F "#{@navi-state}" "send-keys C-k" "select-pane -UZ"
bind-key -n C-l if -F "#{@navi-state}" "send-keys C-l" "select-pane -RZ"
```

**Comment by mrjones2014** [comment]
Woah, this is cool!

Couldn't I theoretically have the plugin set the pane-local tmux option and unset it on nvim exit, so that the background script wouldn't be needed? I actually do something like that for Wezterm already.

https://github.com/mrjones2014/smart-splits.nvim/blob/33c85072ac7901b0f4a68dec7f7d6175f4182377/plugin/smart-splits.vim#L64-L74

This way the default out-of-the-box plugin configuration would have this improvement.

**Reply by b0o** [reply]
Yeah you could do this for sure, I recommend it!

**Reply by b0o** [reply]
It might be a good idea to pick a neutral name for the pane-local option, like `@is-vim`, so that other plugins like `christoomey/vim-tmux-navigator` could choose to adopt it as well.

Edit: I see that the wezterm var name is already plugin-neutral, so you are already doing it that way :)

**Comment by mrjones2014** [comment]
This is now the default implementation since #155, thank you for sharing!
